cobra/core/configuration.py

```diff
@@ -60,7 +65,15 @@ def __init__(self):
                 continue
             else:
                 break
+
+        self.tolerances = {
```

Comment:
I'm not 100% sure about the current behavior in optlang but I know that not all solvers implement all of these tolerances. So I wonder if a simplification to a single tolerance value on the config class would make more sense.

Reply:
Although simplification to a single tolerance value might be an easier to understand design, it might make it a bit difficult if an user wants to access the other tolerance values. A possible solution might be to issue an warning when an user is trying to set a tolerance value which the solver does not possess. I feel we can provide a bit more information about the solvers and their tolerances in the docstring.

Reply:
I agree that they do different things, unfortunately many solvers conflate those. For instance you can not set different tolerances for feasibility and optimality in GLPK. Optlang currently ignores all tolerances except feasibility for GLPK for instance. I think this is something that optlang needs to solve since it is supposed to give a unified interface to different solvers.

Reply:
Well, optlang offers these three settings but some of them don't have an effect for certain solvers. Maybe not the best design but somewhat unified. I wonder if we should go a step further in cobrapy and simply have one tolerance value that is then set on these three optlang attributes.

Reply:
In my opinion, setting a default tolerance value for solvers in cobrapy seems a good idea.

Reply:
So what I'm suggesting is something like:

```python
def __init__(...):
    ...
    self._tolerance = None
    self.tolerance = 1E-09

@property
def tolerance(self):
    return self._tolerance

@tolerance.setter
def tolerance(self, value):
    config = self.solver.Configuration()
    config.tolerances.feasibility = value
    config.tolerances.optimality = value
    config.tolerances.integrality = value
    self._tolerance = value
```

Reply:
Or maybe those properties should actually be on the model class and the configuration holds a simple global default...

Reply:
I can imagine two situations:
- the user is in a notebook-like environment: Using `Configuration` is the recommended way.
- the user is running a program: Is directly setting the solver a better choice? How would using Configuration in that situation be?

I think we should have a separate section on configuring cobrapy internals.

Reply:
I'm thinking that one might want to set the tolerance on a per model basis. Thus there should be a global default in the configuration with which models are initialized but then you should be able to change the parameter for each model. That would mean that `cobra.Configuration.tolerance` is a simple attribute and the properties I outlined above would go onto `cobra.Model.tolerance`.

Reply:
So then `cobra.Model.tolerance` would inherit the value from `cobra.Configuration.tolerance` and `cobra.Model.tolerance` would set the tolerance value for the associated solver. And also the `cobra.Model.tolerance` would allow you to change the value which would in turn change the tolerance for the associated solver. Am I getting it right?

Reply:
Yep, exactly :smiley: